File: monotonicStack/8.bigFishEatSmallFish/py/main.py
# 定义一个函数，参数为n和nums，n为int类型，nums为列表，列表中的元素为int类型
n = int(input())
nums = [int(x) for x in input().split()]
# 定义一个空列表，用于存储栈中的元素
stack = []
# 定义一个变量ans，用于存储最大值
ans = 0
# 遍历nums列表，从最后一个元素开始，每次减1
for i in range(n-1,-1,-1):
    # 定义一个变量cur，用于存储当前元素的最大值
    cur = 0
    # 如果栈不为空，且当前元素大于栈顶元素，则将栈顶元素弹出，并更新cur的值
    while stack and nums[i] > nums[stack[-1]][0]:
        cur = max(cur+1, stack.pop())
    # 将当前元素和cur的值添加到栈中
    stack.append([nums[i],cur])
    # 更新ans的值
    ans = max(cur, ans)
# 输出ans的值
print(ans)

File: monotonicStack/8.bigFishEatSmallFish/go/main.go
package main

import (
	"bufio"
	"os"
	"strconv"
)

var in, out = bufio.NewReader(os.Stdin), bufio.NewWriter(os.Stdout)

func read() int {
	flag := false
	ret := 0
	for c, _ := in.ReadByte(); (c >= '0' && c <= '9') || c == '-'; c, _ = in.ReadByte() {
		if c == '-' {
			flag = true
			continue
		}
		ret = (ret * 10) + int(c-'0')
	}
	if flag {
		ret = -ret
	}
	return ret
}
func main() {
	defer out.Flush()
	n := read()
	nums := make([]int, n)
	for i := 0; i < n; i++ {
		nums[i] = read()
	}
	ans := compute(nums)
	out.WriteString(strconv.Itoa(ans))
	return
}
func compute(nums []int) int {
	n := len(nums)
	stack := make([][2]int, n)
	stackLen := 0
	ans := 0
	for i := n - 1; i >= 0; i-- {
		cur := 0
		// 将前面的小鱼全部吃掉需要多少轮
		for stackLen > 0 && nums[i] > stack[stackLen-1][0] {
			// 吃掉小鱼
			cur = max(cur+1, stack[stackLen-1][1])
			stackLen--
		}
		stack[stackLen] = [2]int{nums[i], cur}
		stackLen++
		ans = max(ans, cur)
	}
	return ans

}
func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
